### PR TITLE
Options for Dry::Types.module

### LINKED ISSUE #289: Options for Dry::Types.module
We want to make dynamic module creation configurable so that the user can choose whether they want their top-level types to be strict/nominal/coercible/etc. Roughly speaking,

```ruby
module Types
  include Dry::Types.module(:strict)
end
```

makes `Types::Integer` to be strict.

We should have done this earlier, now it's scheduled for the 1.0 release.

### PR DESCRIPTION
**Pull request by flash-gordon**
Closes #289

### CODE REVIEW

**Comment by flash-gordon** [comment]
@solnic @timriley let's discuss this, see examples in docs and specs. I'll rename `Definition` to `Nominal` separately

File: lib/dry/types.rb [hunk 1]
```diff
@@ -27,29 +27,55 @@ module Types
...
+    #   # nominal types are exported by default
+    #   Types::Integer
+    #   # => #<Dry::Types[Definition<Integer>]>
+    #   Types::Strict::Integer
+    #   # => #<Dry::Types[Constrained<Definition<Integer> rule=[type?(Integer)]>]>
+    #
+    # @example changing default types
+    #
+    #   module Types
+    #     include Dry::Types(default: :strict)
```
Comment (solnic) [inline lib/dry/types.rb]: There's no `Dry.Types` method, do you intend to add it eventually?

File: lib/dry/types.rb [hunk 2]
```diff
@@ -126,19 +134,13 @@ def self.type_map
       @type_map ||= Concurrent::Map.new
     end
 
-    # List of type keys defined in {Dry::Types.container}
-    # @return [<String>]
-    def self.type_keys
-      container.keys
-    end
-
     private
```
Comment (solnic) [inline lib/dry/types.rb]: This is not needed as it has no effect

File: lib/dry/types.rb (hunk 1 above)
Comment (flash-gordon) [reply lib/dry/types.rb]: yeah, will do it today and merge the whole thing shortly after that

File: lib/dry/types.rb [hunk 3]
```diff
@@ -148,6 +110,55 @@ def self.const_missing(const)
...
     end
   end
+
+  # Export registered types as a module with constants
+  #
+  # @example no options
+  #
+  #   module Types
+  #     # exports all types as constants, uses modules for namespaces
+  #     include Dry::Types.module
```
[the remark this answers is not in the document]
Comment (flash-gordon) [reply lib/dry/types.rb]: right! Thanks for catching it, no idea why I formulated it like this

File: lib/dry/types/module.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          modules.reduce(tree) { |br, name| br[name] ||= {} }[const_name] = type
+        }.freeze
+      end
+
+      private
+
+      # @api private
+      def check_parameters(*namespaces, default: Undefined, **aliases)
+        referenced = namespaces.dup
+        referenced << default unless false.equal?(Undefined.default(default, false))
```
[the remark this answers is not in the document]
Comment (flash-gordon) [reply lib/dry/types/module.rb]: nevermind, I tried to rewrite it several times but at the end of the day it's simply `false.equal?(default) || Undefined.equal?(default)`, I'll change it in master

**Comment by flash-gordon** [comment]
@timriley wrt aliases I don't have an example right now, but since people import constants in _their_ code it makes sense to give them control over the process

**Pull request merged**